File: clojure/plugins/voxgig/sekreto/plugins/onepassword.clj
;; 1Password through a Connect server, as a voxgig/plugin definition.
;;
;; A plugin rather than a built-in kind because it opens a socket
;; (docs/design/plugin-providers.md).
;;
;; A port of typescript/plugins/onepassword.ts, which is canonical.

(ns voxgig.sekreto.plugins.onepassword
  (:require [voxgig.sekreto.core :as core]
            [voxgig.sekreto.addr :refer [checkaddr]]
            [voxgig.sekreto.json :as json]
            [voxgig.sekreto.provider :as provider]
            [voxgig.sekreto.providers :as providers]
            [voxgig.sekreto.plugins.httpjson :as http]))

(defrecord Onepassword [addr token vault vaultid]
  provider/Provider

  ;; 1Password, through a Connect server.
  ;;
  ;; The item titled `api.token` (titles keep their dots), in the named
  ;; vault. The value is the field with purpose PASSWORD, or the field
  ;; labelled `value`. A vault that cannot be found is an error - config
  ;; names it, so its absence is a broken store, not a missing secret.
  (lookup [_ name]
    (core/checkname name)

    (let [useaddr (http/trimslash (or addr ""))
          _ (when (= "" useaddr)
              (throw (core/sekretoerror "sekreto: onepassword: no addr")))
          _ (checkaddr useaddr)

          auth {"authorization" (str "Bearer " (or token ""))}

          id (or @vaultid
                 (reset! vaultid
                         (let [want (or vault "")]
                           (when (= "" want)
                             (throw (core/sekretoerror "sekreto: onepassword: no vault")))

                           (let [res (http/fetchjson "GET" (str useaddr "/v1/vaults") auth)
                                 list (json/asarr (:body res))]
                             (when (or (not= 200 (:status res)) (nil? list))
                               (throw (core/sekretoerror
                                       (str "sekreto: onepassword error: " (:status res)
                                            ": listing vaults"))))

                             (let [found (some (fn [entry]
                                                 (when (or (= want (json/astext (json/dig entry "id")))
                                                           (= want (json/astext (json/dig entry "name"))))
                                                   entry))
                                               list)]
                               (when (nil? found)
                                 (throw (core/sekretoerror
                                         (str "sekreto: onepassword: no vault named " want))))

                               (or (json/astext (json/dig found "id")) ""))))))

          filter (http/uriescape (str "title eq \"" name "\""))
          found (http/fetchjson "GET" (str useaddr "/v1/vaults/" id "/items?filter=" filter) auth)
          items (json/asarr (:body found))]

      (when (or (not= 200 (:status found)) (nil? items))
        (throw (core/sekretoerror
                (str "sekreto: onepassword error: " (:status found) ": finding " name))))

      (when (seq items)
        (let [itemid (or (json/astext (json/dig (first items) "id")) "")
              item (http/fetchjson "GET" (str useaddr "/v1/vaults/" id "/items/" itemid) auth)]

          (when (not= 200 (:status item))
            (throw (core/sekretoerror
                    (str "sekreto: onepassword error: " (:status item) ": reading " name))))

          (let [fields (or (json/asarr (json/dig (:body item) "fields")) [])
                pick (fn [key want]
                       (some (fn [field]
                               (when (= want (json/asstr (json/dig field key))) field))
                             fields))]
            (json/astext (json/dig (or (pick "purpose" "PASSWORD") (pick "label" "value"))
                                   "value")))))))

  (describe [_] (str "onepassword:" (or vault ""))))

(def onepassword
  "The `onepassword` provider kind."
  (providers/providerplugin
   "onepassword"
   (fn [{:keys [addr token vault]}]
     (->Onepassword addr token vault (atom nil)))))

File: lua/src/sekreto/plugins/onepassword.lua
-- The onepassword plugin: 1Password through a Connect server, over
-- HTTPS.
--
-- A port of typescript/plugins/onepassword.ts, which is canonical.

local err = require('sekreto.err')
local name = require('sekreto.name')
local addr = require('sekreto.addr')
local providers = require('sekreto.providers')
local json = require('sekreto.plugins.json')
local httpjson = require('sekreto.plugins.httpjson')
local support = require('sekreto.plugins.support')

local fail = err.fail
local trimslash = providers.trimslash
local providerplugin = providers.providerplugin
local fetchjson = httpjson.fetchjson
local uriescape = support.uriescape

local M = {}

--- 1Password, through a Connect server.
---
--- The item titled `api.token` (titles keep their dots), in the named
--- vault. The value is the field with purpose PASSWORD, or the field
--- labelled `value`. A vault that cannot be found is an error - config
--- names it, so its absence is a broken store, not a missing secret.
local function onepassword(useaddr, token, vault)
  local vaultid = nil

  local function auth()
    return { { 'authorization', 'Bearer ' .. (token or '') } }
  end

  local function resolvevault(address)
    local want = vault or ''
    if '' == want then
      fail('sekreto: onepassword: no vault')
    end

    local res = fetchjson('GET', address .. '/v1/vaults', auth())
    local list = json.asarr(res.body)

    if 200 ~= res.status or nil == list then
      fail('sekreto: onepassword error: ' .. res.status .. ': listing vaults')
    end

    for _, entry in ipairs(list) do
      local id = json.text(json.dig(entry, 'id'))
      if want == id or want == json.text(json.dig(entry, 'name')) then
        return id or ''
      end
    end

    fail('sekreto: onepassword: no vault named ' .. want)
  end

  return {
    lookup = function(secret)
      name.checkname(secret)

      local address = trimslash(useaddr or '')
      if '' == address then
        fail('sekreto: onepassword: no addr')
      end
      addr.checkaddr(address)

      if nil == vaultid then
        vaultid = resolvevault(address)
      end

      local filter = uriescape('title eq "' .. secret .. '"')
      local found = fetchjson(
        'GET', address .. '/v1/vaults/' .. vaultid .. '/items?filter=' .. filter, auth()
      )

      local items = json.asarr(found.body)
      if 200 ~= found.status or nil == items then
        fail('sekreto: onepassword error: ' .. found.status .. ': finding ' .. secret)
      end

      if 0 == #items then
        return nil
      end

      local item = fetchjson(
        'GET',
        address .. '/v1/vaults/' .. vaultid .. '/items/' ..
          (json.text(json.dig(items[1], 'id')) or ''),
        auth()
      )

      if 200 ~= item.status then
        fail('sekreto: onepassword error: ' .. item.status .. ': reading ' .. secret)
      end

      local fields = json.asarr(json.dig(item.body, 'fields')) or {}

      for _, entry in ipairs(fields) do
        if 'PASSWORD' == json.asstr(json.dig(entry, 'purpose')) then
          return json.text(json.dig(entry, 'value'))
        end
      end
      for _, entry in ipairs(fields) do
        if 'value' == json.asstr(json.dig(entry, 'label')) then
          return json.text(json.dig(entry, 'value'))
        end
      end

      return nil
    end,
    describe = function()
      return 'onepassword:' .. (vault or '')
    end,
  }
end

--- The kind, as a voxgig/plugin definition.
M.onepassword = providerplugin('onepassword', function(spec)
  return onepassword(spec.addr, spec.token, spec.vault)
end)

return M
